### PR TITLE
Orca Pytorch cifar10 example bigdl backend

### PR DESCRIPTION
**Pull request by jing-xu**
This is the example of Pytorch Cifar10 tutorial. Follow the original logic, there are sampled graphs showing training data and testing data, trainning on Orca bigdl backend and calculating test accuracy on test data. The part for test accuracy on every class is currently unsupported. The evaluation function in estimator only returns a total accuracy and the predict function in estimator is not implement yet

### CODE REVIEW

**Comment by jing-xu** [comment]
@jason-dai This PR cleaned up code for cifar10. It should follow the original logic, except for the calculation of test accuracy on every class is not supported now.

File: pyzoo/zoo/examples/orca/learn/pytorch/cifar10/cifar10.py [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+    parser.add_argument('--test-batch-size', type=int, default=1000, metavar='N',
+                        help='input batch size for testing per executor(default: 1000)')
+    parser.add_argument('--epochs', type=int, default=2, metavar='N',
+                        help='number of epochs to train (default: 2)')
+    parser.add_argument('--lr', type=float, default=0.001, metavar='LR',
+                        help='learning rate (default: 0.001)')
+    parser.add_argument('--seed', type=int, default=1, metavar='S',
+                        help='random seed (default: 1)')
+    parser.add_argument('--save-model', action='store_true', default=False,
+                        help='For Saving the current Model')
```
Comment (hkvision) [inline pyzoo/zoo/examples/orca/learn/pytorch/cifar10/cifar10.py:78]: remove the unnecessary options, and just give them the same default values as in the tutorial instead.

File: pyzoo/zoo/examples/orca/learn/pytorch/cifar10/cifar10.py [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+    parser.add_argument('--batch-size', type=int, default=4, metavar='N',
+                        help='input batch size for training per executor(default: 256)')
+    parser.add_argument('--test-batch-size', type=int, default=4, metavar='N',
+                        help='input batch size for testing per executor(default: 1000)')
+    parser.add_argument('--epochs', type=int, default=2, metavar='N',
+                        help='number of epochs to train (default: 2)')
+    parser.add_argument('--lr', type=float, default=0.001, metavar='LR',
+                        help='learning rate (default: 0.001)')
+    parser.add_argument('--seed', type=int, default=1, metavar='S',
+                        help='random seed (default: 1)')
```
Comment (hkvision) [inline pyzoo/zoo/examples/orca/learn/pytorch/cifar10/cifar10.py:77]: Can we remove these and just provide default values in the code (not configurable)?

File: pyzoo/zoo/examples/orca/learn/pytorch/cifar10/cifar10.py [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+
+from zoo.orca import init_orca_context, stop_orca_context
+from zoo.orca.learn.pytorch import Estimator
+from zoo.orca.learn.metrics import Accuracy
+from zoo.orca.learn.trigger import EveryEpoch
+
+
+parser = argparse.ArgumentParser(description='PyTorch Cifar10 Example')
+parser.add_argument('--cluster_mode', type=str, default="local",
+                    help='The mode for the Spark cluster. local or yarn.')
```
Comment (jason-dai) [inline pyzoo/zoo/examples/orca/learn/pytorch/cifar10/cifar10.py:41]: help='The cluster mode, such as local, yarn or k8s.'

File: pyzoo/zoo/examples/orca/learn/pytorch/cifar10/cifar10.py [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+               
+dataiter = iter(trainloader)
+images, labels = dataiter.next()
+
+# show images
+imshow(torchvision.utils.make_grid(images))
+# print labels
+print(' '.join('%5s' % classes[labels[j]] for j in range(4)))
+               
+if args.cluster_mode == "local":
```
Comment (jason-dai) [inline pyzoo/zoo/examples/orca/learn/pytorch/cifar10/cifar10.py:75]: move this to the beginning of the code, immediately after argument parsing

File: pyzoo/zoo/examples/orca/learn/pytorch/cifar10/cifar10.py [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+    init_orca_context(cores=1, memory="20g")
```
Comment (jason-dai) [inline pyzoo/zoo/examples/orca/learn/pytorch/cifar10/cifar10.py:76]: do we really need `20g` for memory?

File: pyzoo/zoo/examples/orca/learn/pytorch/cifar10/cifar10.py [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+elif args.cluster_mode == "yarn":
+    init_orca_context(
+        cluster_mode="yarn-client", cores=4, num_nodes=2, memory="2g",
+        driver_memory="10g", driver_cores=1,
```
Comment (jason-dai) [inline pyzoo/zoo/examples/orca/learn/pytorch/cifar10/cifar10.py:80]: do we need 10g memory for driver? And do we really need to specify all the configurations for driver the spark conf?

File: pyzoo/zoo/examples/orca/learn/pytorch/cifar10/cifar10.py [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+        x = F.relu(self.fc2(x))
+        x = self.fc3(x)
+        return x
+   
+net = Net()
+criterion = nn.CrossEntropyLoss()
+optimizer = optim.SGD(net.parameters(), lr=0.001, momentum=0.9)
+
+net.train()  
+zoo_estimator = Estimator.from_torch(model=net, optimizer=optimizer, loss=criterion,
```
Comment (jason-dai) [inline pyzoo/zoo/examples/orca/learn/pytorch/cifar10/cifar10.py:109]: orca_estimator

File: pyzoo/zoo/examples/orca/learn/pytorch/cifar10/cifar10.py [hunk 8]
```diff
@@ -0,0 +1,124 @@
...
+print('Finished Training')
+    
+dataiter = iter(testloader)
+images, labels = dataiter.next()
+
+# print images
+imshow(torchvision.utils.make_grid(images))
+print('GroundTruth: ', ' '.join('%5s' % classes[labels[j]] for j in range(4)))
+
+zoo_estimator.evaluate(data=testloader, validation_methods=[Accuracy()])
```
Comment (jason-dai) [inline pyzoo/zoo/examples/orca/learn/pytorch/cifar10/cifar10.py:122]: add some comments; where is the evaluation result used?

File: pyzoo/zoo/examples/orca/learn/pytorch/cifar10/cifar10.py [hunk 9]
```diff
@@ -0,0 +1,124 @@
...
+               
+if args.cluster_mode == "local":
+    init_orca_context(cores=1, memory="20g")
+elif args.cluster_mode == "yarn":
+    init_orca_context(
+        cluster_mode="yarn-client", cores=4, num_nodes=2, memory="2g",
+        driver_memory="10g", driver_cores=1,
+        conf={"spark.rpc.message.maxSize": "1024",
+            "spark.task.maxFailures": "1",
+            "spark.driver.extraJavaOptions": "-Dbigdl.failure.retryTimes=1"})
```
Comment (hkvision) [inline pyzoo/zoo/examples/orca/learn/pytorch/cifar10/cifar10.py:83]: default value in conf?

File: pyzoo/zoo/examples/orca/learn/pytorch/cifar10/cifar10.py (hunk 6 above)
Comment (hkvision) [reply pyzoo/zoo/examples/orca/learn/pytorch/cifar10/cifar10.py:80]: remove these unnecessary ones

File: pyzoo/zoo/examples/orca/learn/pytorch/cifar10/cifar10.py (hunk 9 above)
Comment (hkvision) [reply pyzoo/zoo/examples/orca/learn/pytorch/cifar10/cifar10.py:83]: Opened an issue https://github.com/intel-analytics/analytics-zoo/issues/3151 for @qiuxin2012 to follow up.

File: pyzoo/zoo/examples/orca/learn/pytorch/cifar10/cifar10.py (hunk 5 above)
Comment (jing-xu) [reply pyzoo/zoo/examples/orca/learn/pytorch/cifar10/cifar10.py:76]: 4g is enough for this case

Comment (hkvision) [reply pyzoo/zoo/examples/orca/learn/pytorch/cifar10/cifar10.py:76]: shall we change default value from 2g to 4g? @jason-dai

Comment (jason-dai) [reply pyzoo/zoo/examples/orca/learn/pytorch/cifar10/cifar10.py:76]: It's OK to pass '4g' as an argument

File: pyzoo/zoo/examples/orca/learn/pytorch/cifar10/cifar10.py (hunk 8 above)
Comment (hkvision) [reply pyzoo/zoo/examples/orca/learn/pytorch/cifar10/cifar10.py:122]: The result is JavaObject and can't be processed from python. Opened an issue: https://github.com/intel-analytics/analytics-zoo/issues/3156

**Comment by hkvision** [comment]
Refer to here: https://github.com/intel-analytics/analytics-zoo/blob/master/pyzoo/zoo/examples/run-example-tests-jep.sh to add the example to our integration test.
Also need to run jenkins.

File: pyzoo/zoo/examples/orca/learn/pytorch/cifar10/cifar10.py [hunk 10]
```diff
@@ -0,0 +1,126 @@
...
+import torch.nn.functional as F
+import torch.optim as optim
+
+from zoo.orca import init_orca_context, stop_orca_context
+from zoo.orca.learn.pytorch import Estimator
+from zoo.orca.learn.metrics import Accuracy
+from zoo.orca.learn.trigger import EveryEpoch
+
+
+os.environ['KMP_DUPLICATE_LIB_OK']='True'
```
Comment (hkvision) [inline pyzoo/zoo/examples/orca/learn/pytorch/cifar10/cifar10.py:39]: why need this?

File: pyzoo/zoo/examples/orca/learn/pytorch/cifar10/cifar10.py [hunk 11]
```diff
@@ -0,0 +1,126 @@
...
+        return x
+   
+net = Net()
+criterion = nn.CrossEntropyLoss()
+optimizer = optim.SGD(net.parameters(), lr=0.001, momentum=0.9)
+
+net.train()  
+orca_estimator = Estimator.from_torch(model=net, optimizer=optimizer, loss=criterion,
+                                     backend="bigdl")
+orca_estimator.fit(data=trainloader, epochs=1, validation_data=testloader,
```
Comment (hkvision) [inline pyzoo/zoo/examples/orca/learn/pytorch/cifar10/cifar10.py:114]: the original script trains 2 epochs?

File: pyzoo/zoo/examples/run-example-tests-jep.sh [hunk 12]
```diff
@@ -3,7 +3,7 @@
 set -e
 
 echo "start example for MNIST"
-#timer
+timer
```
Comment (hkvision) [inline pyzoo/zoo/examples/run-example-tests-jep.sh:6]: ???

File: pyzoo/zoo/examples/orca/learn/pytorch/cifar10/cifar10.py [hunk 13]
```diff
@@ -0,0 +1,126 @@
...
+parser.add_argument('--cluster_mode', type=str, default="local",
+                    help='The cluster mode, such as local, yarn or k8s.')
+args = parser.parse_args()
+
+if args.cluster_mode == "local":
+    init_orca_context(memory="4g")
+elif args.cluster_mode == "yarn":
+    init_orca_context(
+        cluster_mode="yarn-client", cores=4, num_nodes=2, memory="2g",
+        driver_memory="4g", driver_cores=1,
```
Comment (hkvision) [inline pyzoo/zoo/examples/orca/learn/pytorch/cifar10/cifar10.py:51]: no need for driver_memory and driver_cores?

**Comment by hkvision** [comment]
Add back the readme. Others LGTM

File: pyzoo/zoo/examples/orca/learn/pytorch/cifar10/cifar10.py (hunk 10 above)
Comment (hkvision) [reply pyzoo/zoo/examples/orca/learn/pytorch/cifar10/cifar10.py]: do we need this?

File: pyzoo/zoo/examples/orca/learn/pytorch/cifar10/cifar10.py [hunk 14]
```diff
@@ -0,0 +1,125 @@
...
+                   validation_methods=[Accuracy()], checkpoint_trigger=EveryEpoch())
+print('Finished Training')
+dataiter = iter(testloader)
+images, labels = dataiter.next()
+
+# print images
+imshow(torchvision.utils.make_grid(images))
+print('GroundTruth: ', ' '.join('%5s' % classes[labels[j]] for j in range(4)))
+
+result = orca_estimator.evaluate(data=testloader, validation_methods=[Accuracy()])
```
Comment (hkvision) [inline pyzoo/zoo/examples/orca/learn/pytorch/cifar10/cifar10.py:123]: the issue has been fixed, change to get and print the result.

File: pyzoo/zoo/examples/orca/learn/pytorch/cifar10/README.md [hunk 15]
```diff
@@ -0,0 +1,39 @@
...
+We demostrate how to easily run synchronous distributed Pytorch training using Pytorch Estimator of Project Orca in Analytics Zoo. We use a simple convolutional nueral network model to train on Cifar10 dataset, which is a dataset for image classification. See [here](https://pytorch.org/tutorials/beginner/blitz/cifar10_tutorial.html) for the original single-node version of this example provided by Pytorch.
+
+## Prepare environments
+
+We recommend you to use Anaconda to prepare the environments, especially if you want to run on a yarn cluster
+
+```
+conda create -n zoo python=3.7 #zoo is conda environment name, you can set another name you like.
+conda activate zoo
+pip install analytics-zoo==0.9.0.dev0 # or above
```
Comment (hkvision) [inline pyzoo/zoo/examples/orca/learn/pytorch/cifar10/README.md:11]: pip install analytics-zoo  # Needs to install 0.9.0.dev0 or above

File: pyzoo/zoo/examples/orca/learn/pytorch/cifar10/README.md [hunk 16]
```diff
@@ -0,0 +1,39 @@
+# Pytorch Cifar10 example
+We demostrate how to easily run synchronous distributed Pytorch training using Pytorch Estimator of Project Orca in Analytics Zoo. We use a simple convolutional nueral network model to train on Cifar10 dataset, which is a dataset for image classification. See [here](https://pytorch.org/tutorials/beginner/blitz/cifar10_tutorial.html) for the original single-node version of this example provided by Pytorch.
+
+## Prepare environments
+
+We recommend you to use Anaconda to prepare the environments, especially if you want to run on a yarn cluster
+
+```
+conda create -n zoo python=3.7 #zoo is conda environment name, you can set another name you like.
```
Comment (hkvision) [inline pyzoo/zoo/examples/orca/learn/pytorch/cifar10/README.md:9]: Add a space after #

**Review by hkvision — APPROVED** [review]

**Comment by jing-xu** [comment]
http://10.239.47.210:18888/view/ZOO-PR/job/ZOO-PR-Validation/4728/
the failures are irrelevant

**Comment by hkvision** [comment]
http://10.239.47.210:18888/job/ZOO-PR-Pip-JEP-UT-Spark2.4/50/console
Example test passed.

File: pyzoo/zoo/examples/orca/learn/pytorch/cifar10/cifar10.py [hunk 17]
```diff
@@ -0,0 +1,127 @@
...
+              "spark.task.maxFailures": "1",
+              "spark.driver.extraJavaOptions": "-Dbigdl.failure.retryTimes=1"})
+
+transform = transforms.Compose(
+    [transforms.ToTensor(),
+     transforms.Normalize((0.5, 0.5, 0.5), (0.5, 0.5, 0.5))])
+
+trainset = torchvision.datasets.CIFAR10(root='./data', train=True,
+                                        download=True, transform=transform)
+trainloader = torch.utils.data.DataLoader(trainset, batch_size=4,
```
Comment (qiuxin2012) [inline pyzoo/zoo/examples/orca/learn/pytorch/cifar10/cifar10.py:59]: @jason-dai shall we add some parameter(like batch_size and epoch number) to arguments? This exmaple's test will cost 20 mins on jenkins.

Comment (jason-dai) [reply pyzoo/zoo/examples/orca/learn/pytorch/cifar10/cifar10.py:59]: > @jason-dai shall we add some parameter(like batch_size and epoch number) to arguments? This exmaple's test will cost 20 mins on jenkins.

Yes, we can add epoch for testing.
